base/utils/R/read.output.R

```diff
@@ -16,160 +16,241 @@
+##'   dimension.
+##' @param ncfiles Custom character vector of full paths to NetCDF
+##'   files. If `NULL` (default), this list is constructed
+##'   automatically by looking for `YYYY.nc` files in
+##'   `file.path(outdir, runid)`.
+##' @param verbose Logical. If `TRUE`, print status as every year and
+##'   variable is read (default = `FALSE`).
+##' @param print_summary Logical. If `TRUE` (default), calculate and
+##'   print a summary of the means of each variable for each year.
+##' @param start.year,end.year first and last year of output to read.
+##'   Specify as a date-time (only the year portion is used) or as a
+##'   four-digit number or string. If `NA`, reads all years found in
+##'   `outdir`.
+##' @return If `dataframe = FALSE`, a vector of output variables. If
+##'   `dataframe = TRUE`, a `data.frame` of output variables with
+##'   POSIXct timestamps added (`posix` column).
 ##' @export
-##' @author Michael Dietze, David LeBauer
-read.output <- function(runid, outdir, start.year = NA, end.year = NA, variables = "GPP", dataframe= FALSE, pft.name = NULL) {
-  
-  ## vars in units s-1 to be converted to y-1 cflux = c('GPP', 'NPP', 'NEE',
-  ## 'TotalResp', 'AutoResp', 'HeteroResp', 'DOC_flux', 'Fire_flux') # kgC m-2 s-1
+##' @author Michael Dietze, David LeBauer, Alexey Shiklomanov
+read.output <- function(runid, outdir,
+                        start.year = NA,
+                        end.year = NA,
+                        variables = "GPP",
+                        dataframe = FALSE,
+                        pft.name = NULL,
+                        ncfiles = NULL,
+                        verbose = FALSE,
+                        print_summary = TRUE) {
+
+  ## vars in units s-1 to be converted to y-1
+  ## cflux = c('GPP', 'NPP', 'NEE', 'TotalResp', 'AutoResp', 'HeteroResp', 'DOC_flux', 'Fire_flux') # kgC m-2 s-1
   ## wflux = c('Evap', 'TVeg', 'Qs', 'Qsb', 'Rainf') # kgH20 m-2 d-1
-  
-  # create list of *.nc years - look only for files formatted as YYYY.nc, the default pecan output file name standard
-  ncfiles_sub <- list.files(path = outdir, pattern = "^-?[[:digit:]]{4}\\.nc$", full.names = FALSE)
-  ncfiles <- file.path(outdir, ncfiles_sub)
+
+  if ((missing(runid) || missing(outdir)) && is.null(ncfiles)) {
+    PEcAn.logger::logger.severe(
+      "`runid` or `outdir` is missing, and `ncfiles` is NULL.",
+      "Either provide both `runid` and `outdir`, or explicitly specify all `ncfiles`."
+    )
+  }
+
+  # Needed to allow runid and outdir to be missing
+  if (!is.null(ncfiles)) {
+    if (missing(runid)) runid <- NULL
+    if (missing(outdir)) outdir <- NULL
+  }
+
+  # create list of *.nc years - look only for files formatted as
+  # YYYY.nc, the default pecan output file name standard
+  if (is.null(ncfiles)) {
+    ncfiles_sub <- list.files(path = outdir, pattern = "^-?[[:digit:]]{4}\\.nc$", full.names = FALSE)
+    ncfiles <- file.path(outdir, ncfiles_sub)
```

Comment:
How does this differ from `list.files(..., full.names=TRUE)`? I just tried a few test cases and the results were always identical for me.

Reply:
`ncfiles_sub` is used on line 94 below (which I didn't write). More generally, I often find myself using this pattern when I need to extract from either just the filename or just the directory. Whenever I only care about the full path, I agree that `full.names = TRUE` is useful for saving a step.